lib/content-services/src/lib/tree/services/tree.service.ts

```diff
@@ -0,0 +1,127 @@
+/*!
+ * @license
+ * Copyright 2019 Alfresco Software, Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Injectable } from '@angular/core';
+import { DataSource } from '@angular/cdk/collections';
+import { FlatTreeControl } from '@angular/cdk/tree';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { TreeNode } from '../models/tree-node.interface';
+import { TreeResponse } from '../models/tree-response.interface';
+
+@Injectable({ providedIn: 'root' })
+export abstract class TreeService<T extends TreeNode> extends DataSource<T>  {
+    public readonly treeControl: FlatTreeControl<T>;
+    public treeNodesSource = new BehaviorSubject<T[]>([]);
+
+    get treeNodes(): T[] {
+        return this.treeControl.dataNodes;
+    }
+
+    set treeNodes(nodes: T[]) {
+        this.treeControl.dataNodes = nodes;
+        this.treeNodesSource.next(nodes);
+    }
+
+    constructor() {
+        super();
+        this.treeControl = new FlatTreeControl<T>(node => node.level, node => node.hasChildren);
+        this.treeNodes = [];
+    }
+
+    public abstract getSubNodes(parentNodeId: string, skipCount?: number, maxItems?: number): Observable<TreeResponse<T>>;
+
+    /**
+     * Expands node applying subnodes to it.
+     *
+     * @param nodeToExpand Node to be expanded
+     * @param subNodes List of nodes that will be added as children of expanded node
+     */
+    public expandNode(nodeToExpand: T, subNodes: T[]): void {
+        if (nodeToExpand.hasChildren) {
```

Comment:
should probably have some old-fashioned null-check for the parameters :)

Reply:
Added